### PR TITLE
Periodic GC for volume claims

### PR DESCRIPTION
**Pull request by tgross**
Finishes https://github.com/hashicorp/nomad/issues/7825, similar to https://github.com/hashicorp/nomad/pull/7878

This changeset implements a periodic garbage collection of CSI volumes with missing allocations. This can happen in a scenario where a node update fails partially and the allocation updates are written to raft but the evaluations to GC the volumes are dropped. This feature will cover this edge case and ensure that upgrades from 0.11.0 and 0.11.1 get any stray claims cleaned up.

I've tried to mirror the #7878 PR as much as possible here; best reviewed commit-by-commit as well.

### CODE REVIEW

File: command/agent/testdata/basic.hcl [hunk 1]
```diff
@@ -102,32 +102,33 @@ client {
...
+  protocol_version              = 3
+  raft_protocol                 = 3
+  num_schedulers                = 2
+  enabled_schedulers            = ["test"]
+  node_gc_threshold             = "12h"
+  job_gc_interval               = "3m"
+  job_gc_threshold              = "12h"
+  eval_gc_threshold             = "12h"
+  deployment_gc_threshold       = "12h"
+  csi_volume_claim_gc_threshold = "12h"
```
Comment (tgross) [inline command/agent/testdata/basic.hcl:118]: This is the only new line. Curse you golang diffs! 😀

File: command/agent/config_parse_test.go [hunk 2]
```diff
@@ -91,35 +91,36 @@ var basicConfig = &Config{
...
+		ProtocolVersion:           3,
+		RaftProtocol:              3,
+		NumSchedulers:             helper.IntToPtr(2),
+		EnabledSchedulers:         []string{"test"},
+		NodeGCThreshold:           "12h",
+		EvalGCThreshold:           "12h",
+		JobGCInterval:             "3m",
+		JobGCThreshold:            "12h",
+		DeploymentGCThreshold:     "12h",
+		CSIVolumeClaimGCThreshold: "12h",
```
Comment (tgross) [inline command/agent/config_parse_test.go:107]: This is the only new line. Thanks `gofmt`

File: command/agent/testdata/basic.hcl (hunk 1 above)
Comment (langmartin) [reply command/agent/testdata/basic.hcl:118]: but bless you github option to temporarily hide whitespace diffs

File: nomad/core_sched.go [hunk 3]
```diff
@@ -714,33 +717,103 @@ func allocGCEligible(a *structs.Allocation, job *structs.Job, gcTime time.Time,
...
+	c.logger.Debug("CSI volume claim GC scanning before cutoff index",
+		"index", oldThreshold,
+		"csi_volume_claim_gc_threshold", c.srv.config.CSIVolumeClaimGCThreshold)
+
+NEXT_VOLUME:
+	for i := iter.Next(); i != nil; i = iter.Next() {
+		vol := i.(*structs.CSIVolume)
+
+		// Ignore new volumes
+		if vol.CreateIndex > oldThreshold {
```
Comment (langmartin) [inline nomad/core_sched.go:774]: Is the result of testing, or just a good idea? It seems like a pretty good idea either way.

Comment (tgross) [reply nomad/core_sched.go:774]: Inspired by the way we do it for job GC.

**Pull request merged**